Server/Client basic TCP communication

m-kuzmin opened the pull request:
# Changelog
## Server
- Binding to port and accepting connections
- Reading client messages
- Responding to clients
- On <kbd>Ctrl</kbd> + <kbd>C</kbd> closes all connections and stops reading any messages

## Client
- Connecting to server
- Sending messages
- Receiving responces
- In case of an error closes the connection (considers an error to always be the server closing the connection)

m-kuzmin committed:
feat(server): Listen and accept connections

m-kuzmin committed:
feat(client): Connecting to a server

m-kuzmin committed:
feat(client): Send data to server

m-kuzmin committed:
feat(server): Read and display client message

m-kuzmin committed:
feat(server): Send responce after reading

m-kuzmin committed:
feat(client): Read server responce

penkova commented on cmd/client.go:
```diff
@@ -0,0 +1,54 @@
+package main
```
you should have another structure for the project:

cmd:
--client:
----main.go
--server:
----main.go

m-kuzmin replied:
Resolved in https://github.com/m-kuzmin/golang-tcp-client-and-server/pull/1/commits/4717acc9a8c2f3911616dd67ee3f5773f20e3dcb

penkova commented on cmd/server.go:
```diff
@@ -0,0 +1,65 @@
+		log.Fatal(err)
+	}
+	log.Println("Server ready for connections, address:", ln.Addr())
+	defer ln.Close()
```
how about using channels for graceful shutdown?

m-kuzmin replied:
I dont think its applicable here since the Listener should stop when main returns.

m-kuzmin replied:
I think this is resolved now? (8a56c5e)

m-kuzmin committed:
rfct: Moved `main`s into cmd/<name>/main.go

m-kuzmin committed:
lint: Renamed variables

m-kuzmin commented:
Dont merge yet please, I'll add additional graceful shutdown logic to the server.

m-kuzmin committed:
rfct: Made `var`s const; Added package comment

m-kuzmin committed:
feat(server)!: Added graceful shutdown

Added a quit channel and a wait group to stop all goroutines and
shutdown the server gracefuly.

BREAKING CHANGE: `handle_connection` expects two additional args:
`wg WaitGroup` and `quitCh chan struct{}`. This is required for the implementation
of graceful shutdown as these params are used for communication with the
main thread.

m-kuzmin committed:
fix(server): Added `break` to the close loop

m-kuzmin committed:
client: Consider err from conn.Read as Fatal

m-kuzmin commented:
If everything is resolved we can merge now

penkova commented on cmd/server/main.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+func handleConnection(conn net.Conn, wg *sync.WaitGroup, quitCh <-chan struct{}) {
+	address := conn.RemoteAddr()
```
why do you need connection close here?

m-kuzmin committed:
readme: Added currently implemented features
